libraries/testkit/syncgateway.py

```diff
@@ -55,5 +64,211 @@ def restart(self, config):
+        else:
+            data["target"] = "{}".format(target_db)
+        r = requests.post("{}/_replicate".format(self.admin_url), headers=self._headers, data=json.dumps(data))
+        log_request(r)
+        log_response(r)
+        r.raise_for_status()
+
+
+    def get_num_docs(self, db):
+        r = requests.get("{}/{}/_all_docs".format(self.url, db))
+        log_request(r)
+        log_response(r)
+        r.raise_for_status()
+        resp_data = r.json()
+        return resp_data["total_rows"]
+
+    def get_changes_feed_results(self, db):
+        r = requests.get("{}/{}/_changes".format(self.admin_url, db))
+        log_request(r)
+        log_response(r)
+        r.raise_for_status()
+        resp_data = r.json()
+        return resp_data["results"]
+
+    def get_active_tasks(self):
+        r = requests.get("{}/_active_tasks".format(self.admin_url))
+        log_request(r)
+        log_response(r)
+        r.raise_for_status()
+        resp_data = r.json()
+        return resp_data
+
     def __repr__(self):
         return "SyncGateway: {}:{}\n".format(self.hostname, self.ip)
+
+
+
+def wait_until_doc_in_changes_feed(sg, db, doc_id):
+
+    max_tries = 10
+    sleep_retry_seconds = 1
+
+    for attempt in xrange(max_tries):
+        changes_results = sg.get_changes_feed_results(db)
+        for changes_result in changes_results:
+            if changes_result["id"] == doc_id:
+                return
+
+        time.sleep(sleep_retry_seconds)
+
+    raise Exception("Tried to wait until doc {} showed up on changes feed, gave up".format(doc_id))
+
+
+def wait_until_active_tasks_empty(sg):
+
+    max_tries = 10
+    sleep_retry_seconds = 1
+
+    for attempt in xrange(max_tries):
+        active_tasks = sg.get_active_tasks()
+        if len(active_tasks) == 0:
+            return
+        time.sleep(sleep_retry_seconds)
+
+    raise Exception("Tried to wait until _active_tasks were empty, but they were never empty")
+
+def wait_until_active_tasks_non_empty(sg):
+
+    max_tries = 10
+    sleep_retry_seconds = 1
+
+    for attempt in xrange(max_tries):
+        active_tasks = sg.get_active_tasks()
+        if len(active_tasks) > 0:
+            return
+        time.sleep(sleep_retry_seconds)
+
+    raise Exception("Tried to wait until _active_tasks were non-empty, but they were never non-empty")
+
+def wait_until_docs_sync(sg_user, doc_ids):
+    for doc_id in doc_ids:
+        wait_until_doc_sync(sg_user, doc_id)
+
+def wait_until_doc_sync(sg_user, doc_id):
+    max_tries_per_doc = 100
+    sleep_retry_per_doc_seconds = 1
+
+    for attempt in xrange(max_tries_per_doc):
+        try:
+            doc = sg_user.get_doc(doc_id)
+            # if we got a doc, and no exception was thrown, we're done
+            return
+        except:
+            time.sleep(sleep_retry_per_doc_seconds)
+
+    raise Exception("Waited for doc {} to sync, but it never did".format(doc_id))
+
+def assert_does_not_have_doc(sg_user, doc_id):
```

Comment:
Is there a way for this to poll or ensure that this is reliable? For example if doc is PUT to B and is set to replicate to A, it is possible that querying A for the doc may fail if the doc has not yet been replicated. Could this scenario cause a false positive in this case?